packages/core/src/lib/history.unit.test.ts

```diff
@@ -0,0 +1,137 @@
+    expect(getCurrentBranchOrTag).toHaveReturnedWith('main');
+
+    expect(safeCheckout).toHaveBeenCalledTimes(3);
+    // walk commit history
+    expect(safeCheckout).toHaveBeenNthCalledWith(1, 'abc', {
+      forceCleanStatus: undefined,
+    });
+    expect(safeCheckout).toHaveBeenNthCalledWith(2, 'def', {
+      forceCleanStatus: undefined,
+    });
+    // reset
+    expect(safeCheckout).toHaveBeenNthCalledWith(3, 'main', {
+      forceCleanStatus: undefined,
+    });
+  });
+
+  it('should return correct number of results', async () => {
+    const historyOptions: HistoryOptions = {
+      ...(MINIMAL_CONFIG_MOCK as HistoryOptions),
+    };
+
+    const results = await history(historyOptions, ['abc', 'def']);
+
+    expect(results).toHaveLength(2);
+    expect(results).toStrictEqual(['abc-report', 'def-report']);
+  });
+
+  it('should call collect with correct filename and format', async () => {
+    const historyOptions: HistoryOptions = {
+      ...(MINIMAL_CONFIG_MOCK as HistoryOptions),
+    };
+
+    await history(historyOptions, ['abc']);
+    expect(collectAndPersistReports).toHaveBeenCalledTimes(1);
+    expect(collectAndPersistReports).toHaveBeenNthCalledWith(
+      1,
+      expect.objectContaining({
+        persist: expect.objectContaining({
+          filename: 'abc-report',
+          format: ['json'],
+        }),
+      }),
+    );
+  });
+
+  it('should call upload by default', async () => {
+    const historyOptions: HistoryOptions = {
+      ...(MINIMAL_CONFIG_MOCK as HistoryOptions),
+      upload: {
+        server: 'https://server.com/api',
+        project: 'cli',
+        apiKey: '1234',
+        organization: 'code-pushup',
+        timeout: 4000,
+      },
+    };
+    await history(historyOptions, ['abc']);
+
+    expect(upload).toHaveBeenCalledTimes(1);
+    expect(upload).toHaveBeenNthCalledWith(
+      1,
+      expect.objectContaining({
+        persist: expect.objectContaining({ filename: 'abc-report' }),
+      }),
+    );
+  });
+
+  it('should not call upload if uploadReports is set to false', async () => {
+    const historyOptions: HistoryOptions = {
+      ...(MINIMAL_CONFIG_MOCK as HistoryOptions),
+      upload: {
+        server: 'https://server.com/api',
+        project: 'cli',
+        apiKey: '1234',
+        organization: 'code-pushup',
+        timeout: 4000,
+      },
+      uploadReports: false,
+    };
+    await history(historyOptions, ['abc']);
+
+    expect(upload).not.toHaveBeenCalled();
+  });
+
+  it('should not call upload if upload config is not given', async () => {
+    const historyOptions: HistoryOptions = {
+      ...(MINIMAL_CONFIG_MOCK as HistoryOptions),
+    };
+    await history(historyOptions, ['abc']);
+
+    expect(upload).not.toHaveBeenCalled();
+  });
+});
+
+describe('prepareHashes', () => {
+  it('should get all commits from log if no option is passed', () => {
+    expect(
+      prepareHashes({ all: [{ hash: '1' }, { hash: '5' }] }),
```

Comment:
This makes TS unhappy but not sure how to solve that without `as unknown as LogResult`. Any ideas? cc @matejchalk

Reply:
I used real data and implemented the full object